### ISSUE TITLE
Ideas for easier bootstrap of pre-release versions

### ISSUE THREAD

**Issue by lassik** [gambit/gambit#574]
Are the auto-generated `.c` files for gsc good enough for a bootstrap?

If so, could we create zip/tar archives of the `.c` files from CI builds and people could then extract those archives over a Git checkout of Gambit's repo for a faster bootstrap?

The ready-to-run `gsc` binaries from CI artifacts are nice as well, but maybe not as portable as the `.c` files. For example, different Linux distros use different libc implementations (GNU libc, musl libc, etc.) and different `.so` dynamic library versions.

Ping @feeley @daviwil

**Comment by daviwil** [comment]
Oh yeah, I use Guix as my main system and I definitely can't run CI-built Linux binaries :)  Packaged up `.c` files could certainly be interesting, I could easily upload those per-platform as a build artifact.

**Comment by feeley** [comment]
Yes the `.c` files created as part of the bootstrap process on one machine are sufficient for building on another machine.  This is what `make dist` does (a snapshot of all the current `.c` files including the generated ones).  The `.tar.gz` that is produced does not contain the git history, so a `make` in that directory will not trigger the bootstrap process (which is the right thing).

**Comment by lassik** [comment]
So we can use `make dist` more or less as-is and publish the resulting archive files to solve all our bootstrap woes?

The only thing is, does `make dist` generate too many files for our purposese, or is it just the right set?

**Comment by lassik** [comment]
I also tried to run the Linux binaries on Void Linux which uses musl libc, and that didn't work out.

**Comment by daviwil** [comment]
It seems to package up the entire source contents of the repo sans `.git` and any other build output.

**Comment by feeley** [comment]
Yes it contains all the sources necessary, and certainly some more, but it is hard to tell in general what is not essential.  I would say the result of `make dist` is good enough.

**Comment by lassik** [comment]
Could we make a variant `make dist` so there's a sub-target that generates only the right `.c` files? The idea would be to create a `gambit-bootstrap-c.tar.gz` such that we can do something like:

```
curl -O https://github.com/gambit/gambit/releases/download/latest-39735fd/gambit-bootstrap-c.tar.gz
git clone https://github.com/gambit/gambit.git
cd gambit
tar -xzf ../gambit-bootstrap-c.tar.gz

./configure --...
make
make modules
make install
```

and it would go fast :)

**Comment by feeley** [comment]
I just tried, as a test, to copy `*/*.c` and `*/*.h` to an empty directory and tar/gzip that.  This gives a 6 MB .tar.gz and the `make dist` gives a 15 MB .tar.gz, so it is about a 2.5x ratio.  If you include the .scm files the ratio is about 2x.

I'm worried picking the correct set of `.c` and `.h` files is brittle.

Something more robust and interesting (which I've been hoping to get to for many years!) is to have a `gsc` option when compiling that packages all the necessary C sources to build the program on another machine (so it would need to also package a configure script, etc).  That would be appealing to some users who want to develop portable programs but don't want to force their target audience to first install Gambit to build their sofware.

So basically the logic you are talking about would be officially part of the compilation process.

**Comment by daviwil** [comment]
It'd definitely save me a bunch of time, I'm currently bootstrapping Gambit twice (4.9.3 then HEAD) in my private builds, so starting with a pre-bootstrapped compiler would be excellent!

**Comment by lassik** [comment]
> Something more robust and interesting (which I've been hoping to get to for many years!) is to have a gsc option when compiling that packages all the necessary C sources to build the program on another machine (so it would need to also package a configure script, etc). That would be appealing to some users who want to develop portable programs but don't want to force their target audience to first install Gambit to build their sofware.

So gsc would produce `*.[ch]` files and a configure script to reproduce itself?

**Comment by lassik** [comment]
That sounds very neat, but also something that requires a lot of careful work, as you say. What should we use in the meantime to follow Gambit's git master? Does the `make dist` archive overwrite some files it should leave alone if it's extracted over the git master?

Or is the idea to extract the `make dist` archive into `boot/` in the git checkout, instead of extracting it into the root of the git repo? I had that confused.

**Comment by feeley** [comment]
Yes the idea would be to extract the result of `make dist` in a subdirectory (and `boot` would be logically it).  In fact doing it that way would probably work with very minimal extra logic.  The main point is to copy `boot/gsc/gsc` to `gsc-boot` and then `make bootclean;make`.  But that could still fail if the .tar.gz you get is a bit outdated.  To be failproof, you'd need to copy the sources in `gsc` to `boot/gsc` and a make in `boot`, before copying `boot/gsc/gsc` (for the details the logic for this is in the main makefile).

**Comment by feeley** [comment]
I shouldn't say "failproof"... when bootstrapping nothing is failproof... but anyway it would be pretty close.

**Comment by lassik** [comment]
> The main point is to copy `boot/gsc/gsc` to `gsc-boot`

After first running `make` in the `boot/` directory in order to build `boot/gsc/gsc` based on the C sources extracted into `boot/`, right?

>  and then `make bootclean;make`.

In the root of the git checkout, not in `boot/`?

> But that could still fail if the .tar.gz you get is a bit outdated.

We should figure out how to have a stable URL where to get the latest boot sources. GitHub has a very well thought out URL structure - maybe they already provide a magic URL we can use.

> To be failproof, you'd need to copy the sources in `gsc` to `boot/gsc` and a make in `boot`, before copying `boot/gsc/gsc`

Wouldn't this overwrite many of the gsc sources in `boot/` before they have a chance to be built at all?

**Comment by lassik** [comment]
> Yes the idea would be to extract the result of `make dist` in a subdirectory (and `boot` would be logically it).

This is wonderful. Keeping the boot files entirely separate from the main tree ensures we don't mess anything up. Then having extra files in the boot tarball is a minor inconvenience in file size, not a build failure.

**Comment by feeley** [comment]
I'm not sure I understand the problem this is supposed to solve...

Currently, after a `git clone ...;./configure ...` a `make` will trigger the bootstrap process:

1) it uses `git archive ...` and `tar ...` to create the `boot` subdirectory as a copy of the latest release of Gambit and after a `cd boot` it does a `make` of that latest release
2) then it (roughly) does `cp ../gsc/*.scm gsc` (a copy of the compiler sources) and another `make` to create in `gsc/gsc` a Gambit compiler that is built using the latest release but with the compilation logic of HEAD
3) then there is a `cp gsc/gsc ../gsc-boot` to put in place the compiler that will be used to compile the `.scm` files of the main directory
4) after a `cd ..` it does a `make bootclean` to remove all the (stale) .c files
5) finally a `make` creates all the up to date .c files from the .scm files

The .tar.gz that is produced from a `make dist` is a snapshot of the .c files which is normally in sync with the .scm files.  So it can be built with `make` without a bootstrap process.  However it is still a **snapshot** of some bootstrapped commit of Gambit.  So if it is not older than the latest release it could be used in step 1 above instead of the `git archive ...`.  But I see no advantage to do that.

It could be used instead of steps 1 and 2, but only if the snapshot is recent enough (the compiler logic must support all the features that are used in the .scm files from HEAD).  Another way to put it is that between that snapshot and HEAD, there are no changes to the behaviour of the gsc compiler on the .scm files of HEAD.  So I think the problematic part in all of this is to maintain a logical link between HEAD and the required snapshot.

Given that the artifacts are built on every commit it is tempting to simply use the most recent `make dist` snapshot that has been successfully built.  Note that this only works to build HEAD (if you have an old `git clone` the most recent snapshot may be a gsc that is "too new" with features that don't work exactly the same way, or with old features deprecated/removed).  So what is really needed is a way to find, for a particular commit X what is the closest commit Y before X that has a successful build, and it is the snapshot of Y that is needed to replace steps 1 and 2.

BTW, a nasty case where this doesn't work is when **in the same commit** a feature is added to the compiler, and some .scm file(s) in the RTS or other library, depends on that new feature.  This type of commit breaks the bootstrap process and must be solved with additional commits that removes the issue often with a hack that is scheduled to be cleaned up at the next release.  That's why ideally a new feature of the compiler should be added in a commit before using that feature in a subsequent commit.  But human nature as it is, sometimes a bootstrap breaking commit happens and a hack is needed to get out of the mess.

**Comment by lassik** [comment]
> Given that the artifacts are built on every commit it is tempting to simply use the most recent `make dist` snapshot that has been successfully built.

Yes, this would be ideal if it works.

> Note that this only works to build HEAD (if you have an old `git clone` the most recent snapshot may be a gsc that is "too new" with features that don't work exactly the same way, or with old features deprecated/removed).

Yes.

> So what is really needed is a way to find, for a particular commit X what is the closest commit Y before X that has a successful build, and it is the snapshot of Y that is needed to replace steps 1 and 2.

Exactly. I thought of keeping a table matching commit hashes to the filenames of working bootstrap tarballs (it could be automatically generated) but thought it is overengineering.

> BTW, a nasty case where this doesn't work is when **in the same commit** a feature is added to the compiler, and some .scm file(s) in the RTS or other library, depends on that new feature. This type of commit breaks the bootstrap process and must be solved with additional commits that removes the issue often with a hack that is scheduled to be cleaned up at the next release. That's why ideally a new feature of the compiler should be added in a commit before using that feature in a subsequent commit. But human nature as it is, sometimes a bootstrap breaking commit happens and a hack is needed to get out of the mess.

Yes.

Several CI services support writing `[skip ci]` in the commit message to instruct them to skip that commit.

But the best way to avoid this problem is really to disallow direct commits to master and commit everything via pull requests. Then the CI can run on the pull request (and fail there if the PR breaks the above promise). The PR can be fixed before merging it to master. So master will end up with only working commits.

Really this is an excellent example of just the kind of human error that running CI on each pull request was meant to catch :)

**Comment by lassik** [comment]
Another example is if the Gambit repo keeps a file with the commit hash of the latest working bootstrap C files.

The hash would be updated **manually** and the CI build would simply follow orders and check out `boot/` from that commit. Naturally enough, the build would break if that compiler no longer works. A new commit to the hash file would be needed to fix the build.

The result would be that all breaking bootstrap changes would be recorded in Gambit's git log as a result of fixing the builds.

**Comment by lassik** [comment]
It would be really neat if `git log gsc-boot-commit-hash` gave you a log of all the breaking bootstrap changes that way. And you could use the full power of `git log` to explore the context around each breakage.

**Comment by feeley** [comment]
> > Given that the artifacts are built on every commit it is tempting to simply use the most recent `make dist` snapshot that has been successfully built.
> 
> Yes, this would be ideal if it works.
> 
> > Note that this only works to build HEAD (if you have an old `git clone` the most recent snapshot may be a gsc that is "too new" with features that don't work exactly the same way, or with old features deprecated/removed).
> 
> Yes.
> 
> > So what is really needed is a way to find, for a particular commit X what is the closest commit Y before X that has a successful build, and it is the snapshot of Y that is needed to replace steps 1 and 2.
> 
> Exactly. I thought of keeping a table matching commit hashes to the filenames of working bootstrap tarballs (it could be automatically generated) but thought it is overengineering.
> 
> > BTW, a nasty case where this doesn't work is when **in the same commit** a feature is added to the compiler, and some .scm file(s) in the RTS or other library, depends on that new feature. This type of commit breaks the bootstrap process and must be solved with additional commits that removes the issue often with a hack that is scheduled to be cleaned up at the next release. That's why ideally a new feature of the compiler should be added in a commit before using that feature in a subsequent commit. But human nature as it is, sometimes a bootstrap breaking commit happens and a hack is needed to get out of the mess.
> 
> Yes.
> 
> Several CI services support writing `[skip ci]` in the commit message to instruct them to skip that commit.
> 
> But the best way to avoid this problem is really to disallow direct commits to master and commit everything via pull requests. Then the CI can run on the pull request (and fail there if the PR breaks the above promise). The PR can be fixed before merging it to master. So master will end up with only working commits.
> 
> Really this is an excellent example of just the kind of human error that running CI on each pull request was meant to catch :)

I totally agree with PR catching such situations.  Through the evolution of Gambit I have become increasingly disciplined (in particular using PRs instead of direct commits to master), but no one will ever be perfect so human error will always be a factor.  But I do have high hopes that many problematic situations will be avoided.  I remember the times I had to patch the `gsc` **executable** to get out of some bootstrap issue (in those days a `make` would use `gsc/gsc` as the bootstrap compiler and create a new `gsc/gsc`, so even minor things would cause havoc... that's when I started saving backups of the `gsc` executable just in case!).

**Comment by lassik** [comment]
> no one will ever be perfect so human error will always be a factor.

In a GitHub repo's settings it's possible to summarily disallow all non-PR commits to master. If master needs to be fixed manually, it's possible to turn off that setting, push some commits to master and turn it back on again.

Of course, that doesn't stop commits being made to Gambit clones outside GitHub, but it should catch inadvertent mistakes.

> I remember the times I had to patch the `gsc` **executable** to get out of some bootstrap issue

That sounds pretty extreme. Binary patching is always interesting :)

What do you think about manually recording the boot hashes? The more the idea sinks in, the more I like it. You can check out any old commit, and among the checked-out files will be a file that tells you exactly which bootstrap compiler to get.

**Comment by lassik** [comment]
@feeley Since CI is being worked on again, how do you feel about manually recording the gsc boot commit hashes? :)

**Comment by feeley** [comment]
Manually recording which gsc-boot commit is required is both a tedious and error-prone process.  This should be automated.  Ideal would be a script that searches for the most recently produced artifact older than the specific commit being built. But given that this relies on github and the network, there has to be a fallback to the traditional bootstrap method (in particular when the network is down).

Another approach to the problem is to make the gsc source code more immune to the version of Gambit (or even implementation of Scheme) that is used to build gsc-boot.  Anyone interested in building gsc with another R7RS Scheme system to see what changes to the sources are needed?

Also, it would be nice to have a way to determine (at low cost) if a change to the source code of the Gambit RTS or gsc causes bootstrapping problems (CI is not bad, but it has a moderate cost).

Clearly one step in the right direction is to have more frequent releases.  Also, the master branch should be kept "bootable" and I like the idea of disallowing all non-PR commits to master.  Will have to search for this setting...

**Comment by lassik** [comment]
> Manually recording which gsc-boot commit is required is both a tedious and error-prone process.

One could solve this by only updating the hash when a build breaks. It shouldn't break most of the time so not many updates would be needed in practice. This would imply that Gambit is generally bootstrapped using the oldest gsc version that still works for that purpose.

> Another approach to the problem is to make the gsc source code more immune to the version of Gambit (or even implementation of Scheme) that is used to build gsc-boot.

From a user's perspective, this is of course ideal :)

> Anyone interested in building gsc with another R7RS Scheme system to see what changes to the sources are needed?

It would be excellent if gsc only requires R7RS-small in order to be bootstrapped. There are now several portable implementations of the spec. How viable would you guess that to be?

> Also, it would be nice to have a way to determine (at low cost) if a change to the source code of the Gambit RTS or gsc causes bootstrapping problems (CI is not bad, but it has a moderate cost).
>
> Clearly one step in the right direction is to have more frequent releases. Also, the master branch should be kept "bootable" and I like the idea of disallowing all non-PR commits to master. Will have to search for this setting...

All good things if they can be done.

**Comment by feeley** [comment]
> > Anyone interested in building gsc with another R7RS Scheme system to see what changes to the sources are needed?
> 
> It would be excellent if gsc only requires R7RS-small in order to be bootstrapped. There are now several portable implementations of the spec. How viable would you guess that to be?

The part of gsc that is of interest is the Scheme to C source code pipeline (which is needed for the bootstrapping process).  So only the C target is needed to work (the files `gsc/_t-c-*.scm`) plus the front-end of course.

Most dependencies to the Scheme host system are in the file `gsc/_host.scm`.  In principle all dependencies should be in that file but over time some Gambit dependencies have crept into the main Scheme files.  There should be few of these.

One major dependency to Gambit is the reader which reads the source file to compile.  Originally the file `gsc/_source.scm` contained a complete implementation of the reader used by gsc, but it was a pain to keep this in sync with the evolution of the reader in the `read` procedure in the Gambit RTS (`lib/_io.scm`) so now `gsc/_source.scm` simply calls `##read-all-as-a-begin-expr-from-path` defined in `lib/_io.scm`.  There are various ways to "fix" this, the most interesting is probably to develop a R7RS implementation of Gambit's reader (by extracting the appropriate parts of `lib/_io.scm`) and to have both `lib/_io.scm` and `gsc/_source.scm` depend on it.

This reminds me that the following comment in `lib/_io.scm` is no longer valid and should be updated:

    ;; For compatibility between the interpreter and compiler, this section
    ;; must be the same as the corresponding section in the file
    ;; "gsc/_source.scm" (except that ## and ** are exchanged).

**Comment by lassik** [comment]
Having only one reader in the source tree, and making it work on any R7RS system, definitely sounds like a good place to start. Is the RTS reader used as such by gsi, and could the same source file(s) be shared by gsc?

**Comment by feeley** [comment]
It shouldn't be too hard to extract the reader from `lib/_io.scm` and put it in its own file, to be included or imported from `_io.scm`.  The main thing to change is the removal of the `##` namespace prefix (which can be automated with the module system or namespace declaration) and perhaps the `macro-` prefix which is a bit verbose and might cause people to refrain from adopting the code for other purposes.
